File: javascript/085.js
/**
 * @param {character[][]} matrix
 * @return {number}
 */

var largestRectangleArea = function(heights) {
  let stack = [];
  let result = 0;

  heights.forEach(height => {
    if (stack.length === 0 || stack[stack.length - 1] <= height) {
      stack.push(height);
    } else {
      let count = 0;
      while (stack.length > 0 && stack[stack.length - 1] > height) {
        ++count;
        const h = stack.pop();
        if (h * count > result) result = h * count;
      }

      while (count >= 0) {
        stack.push(height);
        --count;
      }
    }
  });

  const length = stack.length;
  for(let i = 0; i < length; ++i) {
    const value = stack[i] * (length - i);
    if (value > result) result = value;
  }

  return result;
};

var maximalRectangle = function(matrix) {
  const rowCount = matrix.length;
  if (rowCount === 0) return 0;
  const colCount = matrix[0].length;
  if (colCount === 0) return 0;

  getHeight = (row, col) => {
    let height = 0;
    while(row >= 0) {
      if (matrix[row][col] === "1") {
        ++height;
        --row;
      } else {
        break;
      }
    }

    return height;
  }

  let maxValue = 0;
  for(let row = 0; row < rowCount; ++row) {
    let heights = [];
    for(let col = 0; col < colCount; ++col) {
      heights.push(getHeight(row, col));
    }
    const largestArea = largestRectangleArea(heights);
    if (largestArea > maxValue) maxValue = largestArea;
  }

  return maxValue;
};

File: ruby/085.rb
def largest_rectangle_area(heights)
  result = 0

  past_heights = []
  heights.each do |height|
    if past_heights.length == 0 || past_heights[-1] <= height
      past_heights.push(height)
    else
      count = 0
      while past_heights.length > 0 && past_heights[-1] > height do
        count += 1
        h = past_heights.pop
        result = h * count if h * count > result
      end

      (count + 1).times do
        past_heights.push(height)
      end
    end
  end

  0.upto(past_heights.length - 1) do |i|
    value = past_heights[i] * (past_heights.length - i)
    result = value if value > result 
  end

  result
end

# @param {Character[][]} matrix
# @return {Integer}
def maximal_rectangle(matrix)
  return 0 if matrix.length == 0
  return 0 if matrix[0].length == 0

  result = 0
  col_count = matrix[0].length
  1.upto(matrix.length) do |row_count|
    heights = []
    0.upto(col_count) do |col|
      values = []
      0.upto(row_count - 1) do |row|
        values.push(matrix[row][col])
      end
      height = 0
      while values.length > 0 && values[-1] == "1"
        height += 1
        values.pop
      end
      heights.push(height)
    end
    sub_result = largest_rectangle_area(heights)
    result = sub_result if sub_result > result
  end

  result
end